checker/src/org/checkerframework/checker/index/samelen/SameLenAnnotatedTypeFactory.java

```diff
@@ -270,15 +279,25 @@ public Void visitNewArray(NewArrayTree node, AnnotatedTypeMirror type) {
                         getAnnotatedType(arrayLength.getExpression())
                                 .getAnnotationInHierarchy(UNKNOWN);
 
-                if (AnnotationUtils.areSameByClass(arrayAnno, SameLenUnknown.class)) {
-                    Receiver rec =
-                            FlowExpressions.internalReprOf(
-                                    this.atypeFactory, arrayLength.getExpression());
-                    if (isReceiverToStringParsable(rec)) {
+                Receiver rec =
+                        FlowExpressions.internalReprOf(
+                                this.atypeFactory, arrayLength.getExpression());
+                if (isReceiverToStringParsable(rec)) {
+                    if (AnnotationUtils.areSameByClass(arrayAnno, SameLenUnknown.class)) {
                         arrayAnno = createSameLen(rec.toString());
+                    } else if (AnnotationUtils.areSameByClass(arrayAnno, SameLen.class)) {
+                        // Ensure that the array whose length is actually being used is part of the
+                        // annotation. If not, add it.
+                        List<String> arrayAnnoArrays =
+                                getValueOfAnnotationWithStringArgument(arrayAnno);
+                        if (!arrayAnnoArrays.contains(rec.toString())) {
+                            arrayAnnoArrays.add(rec.toString());
+                            String[] newArrayAnnoArrays = arrayAnnoArrays.toArray(new String[0]);
+                            Arrays.sort(newArrayAnnoArrays);
```

Comment:
Probably yes. `Arrays.sort`  is also called by `getCombinedSameLen`, which is why I did it here. I'll remove it from both places and add it to `createSameLen` instead.